### DISCUSSION TITLE
Help! [Test Revert error] on testPriceFeedVersionIsAccurate() / Lesson 7

### DISCUSSION THREAD

**Discussion by LareineHan** [Cyfrin/foundry-full-course-cu#857] [Q&A] [answered]
Hi, this is my first question on github. I am on Lesson 7, and I am constantly getting error on 'testPriceFeedVersionIsAccurate()' while testing. 
Obviously, GPT doesn't help so I am here to ask to community! 

## Here comes the error message. 
```zsh
Running 3 tests for test/FundMeTest.t.sol:FundMeTest
[PASS] testMinimumDollarsIsFive() (gas: 9308)
Logs:
  MINIMUM_USD: 5000000000000000000

[PASS] testOwnerIsMsgSender() (gas: 12196)

[FAIL. Reason: EvmError: Revert] testPriceFeedVersionIsAccurate() (gas: 10135)
Traces:
  [10135] FundMeTest::testPriceFeedVersionIsAccurate() 
    ├─ [5099] FundMe::getVersion() [staticcall]
    │   ├─ [0] XXXXXXXXXXXXXXXXXXXXXX (my address) ::version() [staticcall]
    │   │   └─ ← ()
    │   └─ ← "EvmError: Revert"
    └─ ← "EvmError: Revert"

Test result: FAILED. 2 passed; 1 failed; 0 skipped; finished in 2.19s
 
Ran 1 test suites: 2 tests passed, 1 failed, 0 skipped (3 total tests)

Failing tests:
Encountered 1 failing test in test/FundMeTest.t.sol:FundMeTest
[FAIL. Reason: EvmError: Revert] testPriceFeedVersionIsAccurate() (gas: 10135)

Encountered a total of 1 failing tests, 2 tests succeeded
```

I used one of my Metamask testing addresses. 
And the SEPOLIA_RPC_URL is from my Alchemy account. 

and I'll provide all the codes. 

This is my FundMeTest.t.sol 
```solidity
pragma solidity ^0.8.18;

import {Test, console} from "forge-std/Test.sol";
import {FundMe} from "../src/FundMe.sol";
import {DeployFundMe} from "../script/DeployFundMe.s.sol";

contract FundMeTest is Test {
    FundMe fundMe;

    function setUp() external {
        DeployFundMe deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();
    }

    function testMinimumDollarsIsFive() public {
        assertEq(fundMe.MINIMUM_USD(), 5e18); // only works because minimum dollar is starging with 5
        console.log("MINIMUM_USD: %s", fundMe.MINIMUM_USD());
    }

    function testOwnerIsMsgSender() public {
        assertEq(fundMe.i_owner(), msg.sender);
    }

    function testPriceFeedVersionIsAccurate() public {
        uint256 version = fundMe.getVersion();
        assertEq(version, 4);
    }
}
```

and this is my DeployFundMe.s.sol
```solidity
pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {FundMe} from "../src/FundMe.sol";

contract DeployFundMe is Script {
    function run() external returns (FundMe) {
        vm.startBroadcast();
        FundMe fundMe = new FundMe(XXXXXXXXXXXXXXXXXXXXXXXXXXX);
// here goes my metamask test account Sepolia address
        vm.stopBroadcast();
        return fundMe;
    }
}
```

and  FundMe.sol
```solidity
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "lib/chainlink-brownie-contracts/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

error FundMe__NotOwner();

contract FundMe {
    using PriceConverter for uint256;

    mapping(address => uint256) public addressToAmountFunded;
    address[] public funders;

    address public /* immutable */ i_owner;
    uint256 public constant MINIMUM_USD = 5 * 10 ** 18;
    AggregatorV3Interface private s_priceFeed;

    constructor(address priceFeed) {
        i_owner = msg.sender; 
        s_priceFeed = AggregatorV3Interface(priceFeed);
    }

    function fund() public payable {
        require(msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD, "You need to spend more ETH!");
            addressToAmountFunded[msg.sender] += msg.value;
        funders.push(msg.sender);
    }
    
    function getVersion() public view returns (uint256){
        return s_priceFeed.version(); // get the version of s_priceFeed we are passing in.
    }
    
    modifier onlyOwner {
        if (msg.sender != i_owner) revert FundMe__NotOwner();
        _;
    }
    
    function withdraw() public onlyOwner {
        for (uint256 funderIndex=0; funderIndex < funders.length; funderIndex++){
            address funder = funders[funderIndex];
            addressToAmountFunded[funder] = 0;
        }
        funders = new address[](0);

        (bool callSuccess, ) = payable(msg.sender).call{value: address(this).balance}("");
        require(callSuccess, "Call failed");
    }

    fallback() external payable {
        fund();
    }

    receive() external payable {
        fund();
    }

}
```

and lastly, PriceConverter.sol
```solidity
pragma solidity ^0.8.18;

import {AggregatorV3Interface} from "lib/chainlink-brownie-contracts/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";


library PriceConverter {
    function getPrice(
        AggregatorV3Interface priceFeed
    ) internal view returns (uint256) {
        (, int256 answer, , , ) = priceFeed.latestRoundData();

        return uint256(answer * 10000000000);
    }

    function getConversionRate(
        uint256 ethAmount,
        AggregatorV3Interface priceFeed
    ) internal view returns (uint256) {
        uint256 ethPrice = getPrice(priceFeed);
        uint256 ethAmountInUsd = (ethPrice * ethAmount) / 1000000000000000000;
        // the actual ETH/USD conversion rate, after adjusting the extra 0s.
        return ethAmountInUsd;
    }
}
```

**Comment by theirrationalone** [comment]
Please Go to mocks directory and find MockV3Aggregator.sol then Change the version from = 0 to = 4. 

`test/mocks/MockV3Aggregator.sol`

```solidity
contract MockV3Aggregator { 
     // uint256 public constant version = 0; ❌
     uint256 public constant version = 4; ✅
    // .... Remaining Statements/functions
}

It wasn't your fault btw. 

If above mitigation doesn't work?! Please Share your repo if you could. 

;`)

**Reply by LareineHan** [reply]
Thanks for reply! However, I am having trouble finding mocks directory. Where is it located??

**Reply by 0xusmanf** [reply]
You are running this test on Sepolia network and in order to interact with pricefeed there you need to provide a valid ETH/USD price feed address to your `FundMe` contract while deploying it, the right ETH/USD price feed address for sepolia is `0x694AA1769357215DE4FAC081bf1f309aDC325306`. Provide this address instead of `XXXXXXXXXXXXXXXXXXXXXXXXXXX` in your `DeployFundMe.s.sol` where you are creating new `FundMe` contract.

**Reply by LareineHan** [reply]
Thanks for your reply. I already tried with my Ganache address and anvil addresses. However it throw the same error. :(

**Reply by 0xusmanf** [reply, marked as answer]
> 0x694AA1769357215DE4FAC081bf1f309aDC325306

Try it with this address.

Ganache address and anvil addresses are not valid price feed addresses.

**Reply by Rachit2323** [reply]
I add my sepolia address over there but got the error can you please explain why?
Error 
`  [10228] FundMeTest::testPriceVersion()
    ├─ [5121] FundMe::getVersion() [staticcall]
    │   ├─ [0] 0x58187FA71aa989EF1ebe268196463f37FBaAEAfc::version() [staticcall]
    │   │   └─ ← [Stop] 
    │   └─ ← [Revert] EvmError: Revert
    └─ ← [Revert] EvmError: Revert`

**Comment by LareineHan** [comment]
> usmanfarooq91 
Hey thank you. I wonder how '0x694AA1769357215DE4FAC081bf1f309aDC325306 ' this address works. I've tried my Metamask Sepolia testing accounts with Sepolia test funds init but they didn't work. Can you explain please?

**Reply by 0xusmanf** [reply]
Can you specify the command you are using to test it on sepolia. It is working fine for me on sepolia network.

**Reply by LareineHan** [reply]
I've used same command what Patrick used. 
```zsh
forge test -vv --fork-url $SEPOLIA_RPC_URL
```

**Reply by 0xusmanf** [reply]
Working for me with the files you provided. Can you provide the error you are getting?

```shell
usmanfarooq90@DESKTOP-8OOB3I3:~/test$ forge test -vv --fork-url $SEPOLIA_RPC_URL
[⠰] Compiling...
No files changed, compilation skipped

Running 3 tests for test/FundMeTest.t.sol:FundMeTest
[PASS] testMinimumDollarsIsFive() (gas: 9308)
Logs:
  MINIMUM_USD: 5000000000000000000

[PASS] testOwnerIsMsgSender() (gas: 7733)
[PASS] testPriceFeedVersionIsAccurate() (gas: 16161)
Test result: ok. 3 passed; 0 failed; 0 skipped; finished in 4.24s
```

**Reply by LareineHan** [reply]
I mean I used url's from 'https://docs.chain.link/data-feeds/price-feeds/addresses?network=ethereum&page=8' here which is same address you've provided me and Patrick's using. I've used address of my Metamask's Sepolia wallet address and it caused problem. Thank you.

**Comment by LareineHan** [comment]
> Please Go to mocks directory and find MockV3Aggregator.sol then Change the version from = 0 to = 4.
> 
> `test/mocks/MockV3Aggregator.sol`
> 
> ```solidity
> contract MockV3Aggregator { 
>      // uint256 public constant version = 0; ❌
>      uint256 public constant version = 4; ✅
>     // .... Remaining Statements/functions
> }
> 
> It wasn't your fault btw. 
> 
> If above mitigation doesn't work?! Please Share your repo if you could. 
> 
> ;`) 
> ```

Thank you!
Found MockV3Aggregator.sol from
chainlink-brownie-contracts > contracts > src > v0.8 > tests > MockV3Aggregator.sol
and I modified 0 to 4 as you've suggested above. However, it seems like that wasn't the issue here. The address was. I don't know why the other address works but mine.

**Reply by theirrationalone** [reply]
Are you sure you're importing MockV3Aggregator from that location??.. 

According to lesson, patrick made an isolated MockV3Aggregator.sol file for mocking, into test folder -> he created a folder called mock Or mocks whatever, and created MockV3Aggregator.sol file inside there. Please verify if you're also importing mockV3aggregator from that location too. If so, then please modify that mock file as same as you didi above.

**Reply by LareineHan** [reply]
Yeah after I changed the address on DeployFundMe.sol to what Patrick's using (which is sourced from https://docs.chain.link/data-feeds/price-feeds/addresses?network=ethereum&page=8) it finally works. 
After solving this testing problem, I proceeded the lesson and Patrick created 'test/mocks/MockV3Aggregator.sol'. Now I see why he is changing version 0 to 4.  Thank you though. :)
